packages/runtime/src/runtime.ts

```diff
@@ -363,37 +363,63 @@ export class Runtime {
     this.store.globalApps.delete(appId);
   }
 
+  // verify 'amt' microalgos can be withdrawn from account
+  assertMinBalance (amt: number, address: string): void {
+    const account = this.assertAccountDefined(this.store.accounts.get(address));
+    if ((account.amount - amt) < account.minBalance) {
+      throw new TealError(ERRORS.TEAL.INSUFFICIENT_ACCOUNT_BALANCE, {
+        amount: amt,
+        address: address
+      });
+    }
+  }
+
   // transfer ALGO as per transaction parameters
   transferAlgo (txnParam: AlgoTransferParam): void {
     const fromAccount = this.assertAccountDefined(this.store.accounts.get(txnParam.fromAccount.addr));
     const toAccount = this.assertAccountDefined(this.store.accounts.get(txnParam.toAccountAddr));
 
+    this.assertMinBalance(txnParam.amountMicroAlgos, fromAccount.address);
     fromAccount.amount -= txnParam.amountMicroAlgos; // remove 'x' algo from sender
     toAccount.amount += txnParam.amountMicroAlgos; // add 'x' algo to receiver
+
+    if (txnParam.payFlags.closeRemainderTo) {
+      const closeRemToAcc = this.assertAccountDefined(
+        this.store.accounts.get(txnParam.payFlags.closeRemainderTo));
+
+      closeRemToAcc.amount += fromAccount.amount; // transfer funds of sender to closeRemTo account
+      fromAccount.amount = 0; // close sender's account
+    }
   }
 
   /**
    * Update current state and account balances
    * @param txnParams : Transaction parameters
-   * @param accounts : accounts passed by user
    */
-  updateFinalState (txnParams: ExecParams | ExecParams[]): void {
-    let txnParameters;
-    if (!Array.isArray(txnParams)) {
-      txnParameters = [txnParams];
-    } else {
-      txnParameters = txnParams;
-    }
-    for (const txnParam of txnParameters) {
-      switch (txnParam.type) {
-        case TransactionType.TransferAlgo:
-          this.transferAlgo(txnParam);
-          break;
-        case TransactionType.DeleteSSC:
-          this.deleteApp(txnParam.appId);
-          break;
+  updateFinalState (txnParams: ExecParams[]): void {
+    txnParams.forEach((txnParam, idx) => {
+      const fromAccount = this.assertAccountDefined(this.store.accounts.get(txnParam.fromAccount.addr));
+      const fee = this.ctx.gtxs[idx].fee;
+      this.assertMinBalance(fee, txnParam.fromAccount.addr);
+      fromAccount.amount -= fee; // remove tx fee from Sender's account
+
+      if (txnParam.payFlags) {
+        switch (txnParam.type) {
+          case TransactionType.TransferAlgo: {
+            this.transferAlgo(txnParam);
+            break;
+          }
+          case TransactionType.DeleteSSC: {
+            this.deleteApp(txnParam.appId);
+            break;
+          }
+          case TransactionType.CloseSSC || TransactionType.ClearSSC: {
+            fromAccount.closeApp(txnParam.appId); // remove app from local state
```

Comment:
please add a check here for closout - `Close an account out of an application, removing local state from your account. The application must still exist. If it doesn't, use 'goal app clear'.`

Reply:
Ok, so for closeout the application in creator account's global state must exist then? And for clear it will remove the app either way (even if it doesn't exist in creator's globalState)?
Can you share the link to reference?

Reply:
https://developer.algorand.org/docs/reference/cli/goal/app/closeout/#search-overlay

https://developer.algorand.org/docs/reference/cli/goal/app/clear/#search-overlay

Reply:
hmmm ok. It's not mentioned in the smart contract reference docs. Will add links to cli docs as well.